### PR TITLE
Propose enforcing exclusive access to memory

### PR DESCRIPTION
**Pull request by rjmccall**

### CODE REVIEW

File: proposals/NNNN-enforce-exclusive-access-to-memory.md [hunk 1]
```diff
@@ -0,0 +1,587 @@
...
+
+In order to gain the performance and language-design benefits of
+exclusivity, we will have to enforce it in all language modes.
+Therefore, exclusivity will eventually demand a source break.
+
+We can mitigate some of the impact of this break by implicitly migrating
+code matching certain patterns to use different patterns that are known
+to satisfy the exclusivity rule.  For example, it would be straightforward
+to automatically translate calls like ``swap(&array[i], &array[j])`` to
+``array.swap(i, with: j)``.  Whether this makes sense for any particular
```
Comment (JustinJiaDev) [inline proposals/NNNN-enforce-exclusive-access-to-memory.md]: You forgot to change the function name to swapAt.

File: proposals/NNNN-enforce-exclusive-access-to-memory.md [hunk 2]
```diff
@@ -0,0 +1,587 @@
...
+
+Local variables which are captured in closures may sometimes require
+dynamic enforcement instead of static enforcement because the
+implementation cannot locally reason about when the closure will
+be called.  This is especially true when the closure is ``escaping``,
+but can apply even to non-escaping closures in certain circumstances.
+
+In general, Swift is permitted to upgrade dynamic enforcement to
+static enforcement whenever it has sufficient information to do so.
+For example, if the compiler can prove that two acceses to a global
```
Comment (jpsim) [inline proposals/NNNN-enforce-exclusive-access-to-memory.md]: "accesses"

**Comment by rjmccall** [comment]
Thanks, fixed both of those issues.

File: proposals/NNNN-enforce-exclusive-access-to-memory.md [hunk 3]
```diff
@@ -0,0 +1,587 @@
...
+
+In order to gain the performance and language-design benefits of
+exclusivity, we will have to enforce it in all language modes.
+Therefore, exclusivity will eventually demand a source break.
+
+We can mitigate some of the impact of this break by implicitly migrating
+code matching certain patterns to use different patterns that are known
+to satisfy the exclusivity rule.  For example, it would be straightforward
+to automatically translate calls like ``swap(&array[i], &array[j])`` to
+``array.swapAt(i, with: j)``.  Whether this makes sense for any particular
```
Comment (JustinJiaDev) [inline proposals/NNNN-enforce-exclusive-access-to-memory.md]: I think the revised API is without "with" as well?

```array.swapAt(i, j)```

**Pull request merged**